### ISSUE TITLE
FaceVariable does not accumulate properly in parallel

### ISSUE THREAD

**Issue by maguilella** [usnistgov/fipy#856]
A simple test file giving wrong answer (file.py):
```
from numpy import *
from fipy import *

mesh=Gmsh3D('Cylinder.geo')

print shape(mesh.cellCenters.globalValue)
print shape(mesh.faceCenters.globalValue)
```
gives:

(3, 216)
(3, 756)

when running in serial (python file.py), but gives:

(3, 216)
(3, 216)
(3, 545)
(3, 545)

when in parallel (mpirun -np 2 python file.py). 
mesh.cellCenters.globalValue  gives the correct shape in both cases, but mesh.faceCenters.globalValue fails in parallel.

**Comment by maguilella** [comment]
I have just checked that this problem occurs only in fipy 3.4.1, under python 2, but works correctly under fipy 3.4.2 and python3.
Sorry by the noise

**Comment by maguilella** [comment]
Sorry again, but the error is still present in last version and fipy3. This example shows the difference when running in parallel or serial:

# -*- coding: utf-8 -*-
#Test that gives wrong mesh.faceCenters.globalValue when executed in parallel

from numpy import *
from fipy import *

Lx=10;Ly=10;Lz=10
Lxcoarse=6;Lzcoarse=6
nx=ny=nz=1
nxcoarse=nzcoarse=2
xorig=0;yorig=0;zorig=0
dx=1

mesh=Gmsh3D('''
Lx=%(Lx)g;
Ly=%(Ly)g;
Lz=%(Lz)g;

nx=%(nx)g;
ny=%(ny)g;
nz=%(nz)g;

Lcx=%(Lxcoarse)g;
Lcy=%(Lxcoarse)g;
Lcz=%(Lzcoarse)g;

ncx=%(nxcoarse)g;
ncy=%(nxcoarse)g;
ncz=%(nzcoarse)g;

Point(1) = {%(xorig)g, %(yorig)g, %(zorig)g, %(dx)g};
Extrude {Lx, 0, 0} {
  Point{1};Layers{nx};
}

S1[] = Extrude {0, Ly, 0} {
  Line{1};Layers{ny};
};

S2[] = Extrude {Lcx, 0, 0} {
  Line{4};Layers{ncx};
};

S3[] = Extrude {-Lcx, 0, 0} {
  Line{3};Layers{ncx};
};

S4[] = Extrude {0, Lcy, 0} {
  Line{12, 2, 8};Layers{ncy};
};

S5[] = Extrude {0, -Lcy, 0} {
  Line{11, 1, 7};Layers{ncy};
};

allParts[] = {17, 13, 5, 9 ,21, 25, 29, 33, 37};

Transfinite Surface {allParts[]};

Recombine Surface {allParts[]};

Extrude {0, 0, Lz} {
  Surface{allParts[]};Layers{nz};Recombine;
}

Extrude {0, 0, Lcz} {
  Surface{59,147,169, 81, 103, 125, 191,213,235 };Layers{ncz};Recombine;
}
Extrude {0, 0, -Lcz} {
  Surface{17, 21, 25, 29, 33, 37, 9, 5, 13};Layers{ncz};Recombine;
}
Physical Volume("Volume") = {19, 20, 21, 9, 3, 4, 10, 11, 12, 25, 26, 27, 5, 1, 2, 13, 14, 15, 16, 17, 18, 8, 6, 7, 22, 23, 24};
''' % locals())

print(shape(mesh.cellCenters.globalValue))
print(shape(mesh.faceCenters.globalValue))

**Comment by guyer** [comment]
This is a longstanding issue (#400). The correct answer is for us to remove `FaceVariable.globalValue`, but that's not trivial to do and it hasn't gotten attention.

What's your use case? We may be able to help with an alternative.

**Comment by maguilella** [comment]
Hi Jonathan, thank you for your prompt answer

I'm using it to select surfaces at specific z, in order to compute 
fluxes. For example as in this code:

/zleft=unique(mesh.faceCenters.globalValue[2]*isclose(abs(mesh_faceNormals.globalValue[2]),1))[1]/

I think that the correct way would be to define physical surfaces via 
gmsh and then use it to compute fluxes. Would this work also in 
parallel?? Do you have any other suggestion??

Marcel


El 22/4/22 a les 23:43, Jonathan Guyer ha escrit:

**Comment by guyer** [comment]
Yes, the correct way to designate the faces you want is with Gmsh physical surfaces and, yes, it works in parallel. Even if you were to define the faces parametrically, as in your example, you would want to use `.value`, not `.globalValue`.

**Comment by maguilella** [comment]
Thank you Jonathan

I'm still confused. I have used this example 
(https://gist.github.com/guyer/c46051ddb7c4ad9fdf80ac57205945b1) to add 
a physical surface, as directed. The result is this:

# -*- coding: utf-8 -*-

from numpy import *
from fipy import *


#https://gist.github.com/guyer/c46051ddb7c4ad9fdf80ac57205945b1
geo = """
cellSize = %(cellSize)g;

Point(1) = {0, 0, 0, cellSize};
Point(2) = {4, 0, 0, cellSize};
Point(3) = {20, 0, 0, cellSize};
Point(4) = {20, 10, 0, cellSize};
Point(5) = {4, 10, 0, cellSize};
Point(6) = {0, 10, 0, cellSize};

Point(11) = {0, 0, 10, cellSize};
Point(12) = {4, 0, 10, cellSize};
Point(13) = {20, 0, 10, cellSize};
Point(14) = {20, 10, 10, cellSize};
Point(15) = {4, 10, 10, cellSize};
Point(16) = {0, 10, 10, cellSize};

Line(1) = {1, 2};
Line(2) = {2, 3};
Line(3) = {3, 4};
Line(4) = {4, 5};
Line(5) = {5, 6};
Line(6) = {6, 1};
Line(7) = {2, 5};

Line(13) = {4, 14};
Line(16) = {16, 6};
Line(17) = {5, 15};

Line(21) = {11, 12};
Line(22) = {12, 13};
Line(23) = {13, 14};
Line(24) = {14, 15};
Line(25) = {15, 16};
Line(26) = {16, 11};
Line(27) = {12, 15};

Line(33) = {3, 13};
Line(36) = {11, 1};
Line(37) = {2, 12};

Line Loop(1) = {1, 7, 5, 6};
Line Loop(2) = {2, 3, 4, -7};

Line Loop(11) = {-5, 17, 25, 16};
Line Loop(12) = {-4, 13, 24, -17};

Line Loop(21) = {21, 27, 25, 26};
Line Loop(22) = {22, 23, 24, -27};

Line Loop(31) = {1, 37, -21, 36};
Line Loop(32) = {2, 33, -22, -37};

Line Loop(40) = {-6, -16, 26, 36};
Line Loop(50) = {7, 17, -27, -37};
Line Loop(60) = {3, 13, -23, -33};

Plane Surface(1) = {1};
Plane Surface(2) = {2};

Plane Surface(11) = {11};
Plane Surface(12) = {12};

Plane Surface(21) = {21};
Plane Surface(22) = {22};

Plane Surface(31) = {31};
Plane Surface(32) = {32};

Plane Surface(40) = {40};
Plane Surface(50) = {50};
Plane Surface(60) = {60};

Surface Loop(1) = {1, 11, 21, 31, 40, 50};
Surface Loop(2) = {2, 12, 22, 32, 50, 60};

Volume(1) = {1};
Volume(2) = {2};

Physical Surface("plane") = {50};

Physical Volume("box") = {1,2};
"""

mesh=Gmsh3D(geo % {"cellSize": 1.})


phi=CellVariable(mesh=mesh,value=mesh.z)

Current=( 
phi.faceGrad.dot([0,0,1])*mesh.physicalFaces["plane"]*mesh._faceAreas 
).sum()

print(Current)


I this example I compute de total flux (Current) of phi.faceGrad through 
the physical faces "plane" and still get different (incorrect) values 
when run in parallel (p.e. mpirun -np 4....etc) compared when I run in 
serial (python prova.py). The problem is not solved by using "
( phi.faceGrad.dot([0,0,1])*mesh.physicalFaces["plane"]*mesh._faceAreas 
).globalValue.sum()"

What I'm doing wrong?

Marcel


El 27/4/22 a les 15:17, Jonathan Guyer ha escrit:

**Comment by guyer** [comment]
You weren't doing anything wrong. We apparently never accounted for Gmsh ghost faces. In general, this doesn't matter, but it's important for reduction operations like `.sum()`. Thanks for the good example to diagnose.

I think I've corrected the issue with #857.

Note that
```python
Current = (phi.faceGrad.dot(mesh._orientedAreaProjections)*mesh.physicalFaces["plane"]).sum()
```
makes fewer assumptions than what you wrote.

**Comment by guyer** [comment]
Note that `FaceVariable.globalValue` still doesn't work and likely won't ever.

**Comment by maguilella** [comment]
El 30/4/22 a les 2:29, Jonathan Guyer ha escrit:
>
> You weren't doing anything wrong. We apparently never accounted for 
> Gmsh ghost faces. In general, this doesn't matter, but it's important 
> for reduction operations like |.sum()|. Thanks for the good example to 
> diagnose.
>
OK. Thank you for your great support
>
> I think I've corrected the issue with #857 
> <https://github.com/usnistgov/fipy/pull/857>.
>
I'll check asap: Will it be included in the last version?
>
> Note that
>
> Current  =  (phi.faceGrad.dot(mesh._orientedAreaProjections)*mesh.physicalFaces["plane"]).sum()
>
> makes fewer assumptions than what you wrote.
>
Thank you. I didn't know about mesh._orientedAreaProjections...


—
Reply to this email directly, view it on GitHub 
<https://github.com/usnistgov/fipy/issues/856#issuecomment-1113874937>, 
or unsubscribe

**Comment by maguilella** [comment]
I'll keep it in mind.

Thank you!

El 30/4/22 a les 2:31, Jonathan Guyer ha escrit:

**Comment by guyer** [comment]
> I'll check asap: Will it be included in the last version?

Yes

> Thank you. I didn't know about mesh._orientedAreaProjections...

There's no reason you should. The leading `_` means that it's an internal field that we didn't think users needed to worry about. There are several things like this that it's become clear that we need to make them public and document them.

**Comment by guyer** [comment]
In the course of correcting the face accumulation issue, and a variety of edge cases, I realized that it is possible to fix `FaceVariable.globalValue` after all (#400, #857)

**Comment by maguilella** [comment]
Thank you Jonathan for your work. Does this corrected version instal with conda update -c conda-forge fipy?? Or should I download the github version and install manually?

**Comment by guyer** [comment]
Updating the conda-forge recipe is my next order of business (which I was about to forget about, so thank you for the reminder). It should hopefully work quite soon, but for the moment, you'll need to install manually.

**Comment by guyer** [comment]
Unfortunately, [anaconda](https://github.com/conda/infra/issues/561) and/or [conda-forge](https://github.com/conda-forge/conda-forge.github.io/issues/1779) are having a major breakage right now, leading to 3.4.3 not being available via `conda`, yet.

**Comment by maguilella** [comment]
Seems they solved the problem: Have just updated to 3.4.3. I'll let you know how it goes.
Thank you!
